File: Chapter 03/C#/Section_3_2_Coupling_Exercise/Before/UserReportGenerator.cs
using System.Collections.Generic;

namespace Chapter03.CouplingTest.Before;

/// <summary>
/// Generates a report for a user.
/// This class is an example of **High Coupling**.
/// </summary>
public class UserReportGenerator
{
    private readonly UserDataService _dataService = new();

    public string GenerateReport(int userId)
    {
        // PROBLEM: High Coupling and "Chatty" Communication.
        // This report generator is "too smart." It knows the intimate details
        // of how to build a user report. It has to call multiple, fine-grained
        // methods on the data service to get individual pieces of data.
        // This creates a tight bond between this class and the UserDataService.
        string name = _dataService.GetUserName(userId);
        string email = _dataService.GetUserEmail(userId);
        List<string> orders = _dataService.GetUserOrderIds(userId); 

        // PROBLEM: Misplaced Responsibility.
        // The logic for calculating the total spent for a user lives here, inside the
        // client. What if another part of the application also needs to calculate this?
        // That logic would have to be duplicated. The data service should be responsible
        // for providing this calculation.
        decimal totalSpent = 0;
        foreach (var orderId in orders)
        {
            totalSpent += _dataService.GetOrderTotal(orderId);
        }

        // Because of the high coupling, this class is fragile. If the UserDataService
        // changes—for example, if GetUserOrderIds starts returning a different type—this
        // class will break and need to be updated.
        return $"User Report for {name} ({email}) - Total Spent: ${totalSpent}";
    }
}

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_2_coupling_exercise/before/UserReportGenerator.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_2_coupling_exercise.before;

import java.util.List;

/**
 * ANTI-PATTERN: The Tightly Coupled Client.
 * * ARCHITECTURE PROBLEM: Because the UserDataService is too granular, this client 
 * class is forced to take on the responsibility of orchestrating the data gathering. 
 * * The generator now knows WAY too much about the internal structure of the data. 
 * It knows that it has to fetch the user, then fetch the orders, and then loop 
 * through the orders to calculate a total. This is a severe violation of encapsulation.
 */
public class UserReportGenerator {
    private final UserDataService dataService = new UserDataService();

    public String generateReport(int userId) {
        
        // 🚨 ARCHITECTURE WARNING: High temporal coupling. The client has to 
        // call these specific methods in a specific sequence to get what it needs.
        String name = dataService.getUserName(userId);
        String email = dataService.getUserEmail(userId);
        List<String> orders = dataService.getUserOrderIds(userId);

        double totalSpent = 0.0;
        
        // 🚨 ARCHITECTURE WARNING: Feature Envy / Chatty Execution.
        // The client is forced to loop through orders and request totals one by one.
        // It is doing the heavy lifting that the service should be doing for it!
        for (String orderId : orders) {
            totalSpent += dataService.getOrderTotal(orderId);
        }

        return String.format("User Report for %s (%s) - Total Spent: $%.2f", name, email, totalSpent);
    }
}